Mathlib/Dynamics/SymbolicDynamics/Basic.lean

```diff
@@ -0,0 +1,746 @@
+
+Shift invariance guarantees that the ambient shift restricts to `Y`.
+
+**Ergonomics.**
+Thin wrappers (e.g. `Subshift.shift`, `Subshift.cylinder`, `Subshift.languageOn`)
+may be added for convenience. They introduce no new theory and unfold to the
+ambient definitions.
+
+## Namespacing policy
+
+All ambient definitions live under the namespace `SymbolicDynamics.FullShift`.
+If inner, subshift-relative wrappers are provided, they will be placed in the
+subnamespace `SymbolicDynamics.Subshift`. This separation avoids name clashes
+between the two viewpoints, since both may naturally want to reuse names like
+`cylinder`, `shift`, `occursAt`, or `languageOn`.
+
+## Implementation notes
+
+* Openness/closedness results for cylinders and occurrence sets use
+  `[DiscreteTopology A]`. Closedness proofs that enumerate values additionally
+  require `[Fintype A]`, `[DecidableEq A]`, and `[DecidableEq G]` (for `Finset`
+  manipulations and `Function.update`).
+-/
+
+
+noncomputable section
+open Set Topology
+
+namespace SymbolicDynamics
+
+namespace FullShift
+
+/-! ## Full shift and shift action -/
+
+section ShiftDefinition
+
+variable {A G : Type*} [Monoid G]
+
+/-- The **right-translation shift** on configurations.
+
+We call *configuration* an element of `G → A`.
+
+Given a configuration `x : G → A` and an element `g : G` of the group, the shifted configuration
+`mulShift g x` is defined by `(mulShift g x) h = x (h * g)`.
+
+Intuitively, this moves the whole configuration "in the direction of `g`": the value
+at position `h` in the shifted configuration is the value that was at position
+`h * g` in the original one.
+
+For example, if `G = ℤ` (with addition) and `A = {0,1}`, then
+`mulShift 1 x` is the sequence obtained from `x` by shifting every symbol one
+step to the left. -/
+@[to_additive shift
+/-- The **right-translation shift** on configurations, in additive notation.
+
+We call *configuration* an element of `G → A`.
+
+Given a configuration `x : G → A` and an element `g : G` of the multiplicative group,
+the shifted configuration `shift g x` is defined by `(shift g x) h = x (h + g)`.
+
+Intuitively, this moves the whole configuration "in the direction of `g`": the value
+at position `h` in the shifted configuration is the value that was at position
+`h + g` in the original one.
+
+For example, if `G = ℤ` and `A = {0,1}`, then
+`shift 1 x` is the sequence obtained from `x` by shifting every symbol one
+step to the left. -/]
+def mulShift (g : G) (x : G → A) : G → A :=
+  fun h => x (h * g)
+
+@[to_additive] lemma mulShift_apply (g h : G) (x : G → A) :
+  mulShift g x h = x (h * g) := rfl
+
+@[to_additive] lemma mulShift_one (x : G → A) : mulShift (1 : G) x = x := by
+  ext h; simp [mulShift]
+
+/-- Composition of right-translation shifts corresponds to multiplication in the group. -/
+@[to_additive shift_mul] lemma mulShift_mul (g₁ g₂ : G) (x : G → A) :
```

Comment:
why `shift_mul`? I'd rather say `shift_add`. But normally this should be automatic, no need to specify the name by hand.